clients/cobol-lsp-vscode-extension/src/commands/SmartTabCommand.ts

```diff
@@ -72,25 +87,60 @@ class SmartOutdentCommandProvider extends SmartCommandProvider {
         for (let selection of editor.selections) {
             const position = selection.active;
     
-            const firstPosition = this.findFirstPosition(editor, position.line);
-            const prevPosition = getPrevPosition(editor, firstPosition);
-
-            edit.delete(new vscode.Range(new vscode.Position(position.line, prevPosition), new vscode.Position(position.line, firstPosition)));
+            let charPosition = this.findSolidCharPosition(editor, position);
+            if (charPosition == 0) {
+                return;
+            }
+            let prevPosition = getPrevPosition(editor, charPosition);
+            if (this.onlySpaces(getCurrentLine(editor, position.line), prevPosition, charPosition)) {
+                edit.delete(new vscode.Range(new vscode.Position(position.line, prevPosition), new vscode.Position(position.line, charPosition)));
+            } else {
+                let prevSolidPosition = this.findPrevSolidPosition(editor, new vscode.Position(position.line, charPosition - 1));
+                let removeSize = Math.max(0, Math.min(charPosition - prevSolidPosition - 1, TAB_SPACES));
+                edit.delete(new vscode.Range(new vscode.Position(position.line, charPosition - removeSize), new vscode.Position(position.line, charPosition)));
+                prevPosition = charPosition - removeSize;
+            }
             const newPosition: vscode.Position = new vscode.Position(position.line, prevPosition);
             newSelections.push(new vscode.Selection(newPosition, newPosition));
         }
         editor.selections = newSelections;    
     }
 
-    private findFirstPosition(editor: vscode.TextEditor, line: number): number {
-        const textLine: vscode.TextLine = editor.document.lineAt(line);
-        const text = textLine.text;
-        for (let i = 0; i < text.length; i++) {
+    private findSolidCharPosition(editor: vscode.TextEditor, position: vscode.Position) {
+        let result = 0;
+        if (this.isCurrentSpace(editor, position)) {
+            result = findNextSolidPosition(editor, position);
+        } else {
+            result = this.findPrevSolidPosition(editor, position);
```

Comment:
Seems like `this` reference is not needed here

Reply:
`findNextSolidPosition ` is defined in the module scope, but `findPrevSolidPosition ` in the class scope, so it is needed here :-(